src/customFileParser/unpackGeometry/Cylinder.ts

```diff
@@ -1,32 +1,218 @@
-import { RenderedGeometryGroups } from '../sharedFileParserTypes';
 import PropertyLoader from '../PropertyLoader';
 import * as THREE from 'three';
-import { xAxis, zAxis } from './../../constants';
+import { xAxis, yAxis, zAxis } from './../../constants';
+import { addOpenCone, addClosedCone } from './Cone';
+import { truncate } from 'fs';
 
 const centerA = new THREE.Vector3();
 const centerB = new THREE.Vector3();
+const extA = new THREE.Vector3();
+const extB = new THREE.Vector3();
 
+const axisRotation = new THREE.Quaternion();
 const localXAxis = new THREE.Vector3();
+const slicingPlaneNormal = new THREE.Vector3();
+const globalVertex = new THREE.Vector3();
 const rotation = new THREE.Quaternion();
+const globalPlanes = [new THREE.Plane(), new THREE.Plane()];
+const globalLine = new THREE.Line3();
+const globalLineStart = new THREE.Vector3();
+const globalLineEnd = new THREE.Vector3();
+const globalVector = new THREE.Vector3();
 
-function addClosedCylinder(groups: RenderedGeometryGroups, data: PropertyLoader) {
+const globalCapZAxis = new THREE.Vector3();
+const globalCapXAxis = new THREE.Vector3();
+const globalVertices = [new THREE.Vector3(), new THREE.Vector3(), new THREE.Vector3(), new THREE.Vector3()];
+
+export function addClosedCylinder(groups: {[name: string]: any}, data: PropertyLoader) {
   centerA.copy(data.normal).multiplyScalar(data.height / 2).add(data.center);
   centerB.copy(data.normal).multiplyScalar(-data.height / 2).add(data.center);
-  groups.Cone.add(data.nodeId, data.treeIndex, data.color, centerA, centerB,
+  groups.Cone.add(data.nodeId, data.treeIndex, centerA, centerB,
     data.radiusA, data.radiusA, data.rotationAngle);
-  groups.Circle.add(data.nodeId, data.treeIndex, data.color, centerA, data.normal, data.radiusA);
-  groups.Circle.add(data.nodeId, data.treeIndex, data.color, centerB, data.normal, data.radiusA);
+  groups.Circle.add(data.nodeId, data.treeIndex, centerA, data.normal, data.radiusA);
+  groups.Circle.add(data.nodeId, data.treeIndex, centerB, data.normal, data.radiusA);
+}
+
+export function addOpenCylinder(groups: {[name: string]: any}, data: PropertyLoader) {
+  centerA.copy(data.normal).multiplyScalar(data.height / 2).add(data.center);
+  centerB.copy(data.normal).multiplyScalar(-data.height / 2).add(data.center);
+  groups.Cone.add(data.nodeId, data.treeIndex, centerA, centerB,
+  data.radiusA, data.radiusA, data.rotationAngle);
+}
+
+export function addOpenGeneralCylinder(groups: {[name: string]: any}, data: PropertyLoader) {
+  if (data.radiusA !== data.radiusB) {
+    drawHollowedCone(groups, data, false);
+  } else {
+    drawGeneralCylinders(groups, data, false);
+  }
 }
 
-function addGeneralCylinder(groups: RenderedGeometryGroups, data: PropertyLoader) {
-  // TODO
+export function addClosedGeneralCylinder(groups: {[name: string]: any}, data: PropertyLoader) {
+  if (data.radiusA !== data.radiusB) {
+    drawHollowedCone(groups, data, true);
```

Comment:
What does `draw` mean here? Seems like we are rendering it somewhere. Maybe it should be `add`?